regexec.c

```diff
@@ -4218,7 +4218,8 @@ match_at(regex_t* reg, const UChar* str, const UChar* end,
 
  timeout:
   xfree(xmalloc_base);
-  xfree(stk_base);
+  if (stk_base != stk_alloc || IS_NOT_NULL(msa->stack_p))
```

Comment:
Why do we need to check that `stk_base != stk_alloc` here?

Reply:
https://github.com/ruby/ruby/blob/dde99215f2bc60c22a00fc941ff7f714f011e920/regexec.c#L1177-L1181

It looks to me like if `stk_base == stk_alloc` then it could be stack allocated?

But maybe we don't need this check as long as we do a `STACK_SAVE` first? I'm not sure if we do that though. Still getting my bearings in this code.

Reply:
Confirmed that if we only check `IS_NOT_NULL(msa->stack_p)` then we end up with a memory leak again.

But if we call `STACK_SAVE` first then it works. Maybe we could even do:

```
  STACK_SAVE;
  xfree(xmalloc_base);
  xfree(msa->stack_p);
```

What do you think?